nwb_conversion_tools/utils/spike_interface.py

```diff
@@ -331,27 +332,65 @@ def add_electrodes(
+
+    if 'Ecephys' not in metadata:
+        metadata['Ecephys'] = dict()
+
     if 'Electrodes' not in metadata['Ecephys']:
         metadata['Ecephys']['Electrodes'] = []
 
     assert all([isinstance(x, dict) and set(x.keys()) == set(['name', 'description', 'data'])
                 and isinstance(x['data'], list) for x in metadata['Ecephys']['Electrodes']]), \
         "Expected metadata['Ecephys']['Electrodes'] to be a list of dictionaries!"
     assert all([x['name'] != 'group' for x in metadata['Ecephys']['Electrodes']]), \
-        "Passing metadata field 'group' is depricated; pass group_name instead!"
+        "Passing metadata field 'group' is deprecated; pass group_name instead!"
 
     if nwbfile.electrodes is None:
         nwb_elec_ids = []
     else:
         nwb_elec_ids = nwbfile.electrodes.id.data[:]
 
-    for metadata_column in metadata['Ecephys']['Electrodes']:
-        if (nwbfile.electrodes is None or metadata_column['name'] not in nwbfile.electrodes.colnames) \
-                and metadata_column['name'] not in defaults:
-            nwbfile.add_electrode_column(
-                name=str(metadata_column['name']),
-                description=str(metadata_column['description'])
-            )
+    elec_columns = {}  # name: description
+    property_names = set()
+    for chan_id in recording.get_channel_ids():
+        for i in recording.get_channel_property_names(channel_id=chan_id):
+            property_names.add(i)
+    for prop in property_names:
+        if prop not in ['gain', 'offset', 'location', 'name'] + list(exclude):
+            # property 'gain' should not be in the NWB electrodes_table
+            # property 'brain_area' of RX channels corresponds to 'location' of NWB electrodes
+            # property 'offset' should not be in the NWB electrodes_table as not officially supported by schema v2.2.5
+
+            data = []
+            for chan_id in recording.get_channel_ids():
+                if prop in recording.get_channel_property_names(channel_id=chan_id):
+                    data.append(recording.get_channel_property(channel_id=chan_id, property_name=prop))
+                else:
+                    if len(data) > 0 and isinstance(data[-1], str):
+                        data.append('')
+                    elif len(data) > 0:
+                        data.append(np.nan)
+            prop = 'location' if prop == 'brain_area' else prop
+            prop = 'group_name' if prop == 'group' else prop
+            elec_columns[prop] = dict(description=prop, data=data)
+
+    for x in metadata['Ecephys']['Electrodes']:
+        name = x.pop('name')
+        if name not in elec_columns:
+            elec_columns[name] = dict(description=name)
+        if 'description' in x:
+            elec_columns[name]['description'] = x['description']
+        if 'data' in x:
+            assert len(x['data']) == recording.get_num_channels()
+            elec_columns[name]['data'] = x['data']
```

Comment:
Note that you don't have to test for the conditional existence of these fields. The format assertion from [these lines ](https://github.com/catalystneuro/nwb-conversion-tools/pull/181/files#diff-296417a9c87b534b7bc88ddf071347b6a0d5241391a80448db87082db15c55b5R342-R344)ensures they exist so you can pull from them directly.

Reply:
Well first of all, those lines should be replaced by just validating json-schema. Second of all, I'd like to relax this constraint and allow for 'data' to not be defined in the metadata dict.

Reply:
I've pushed [changes here ](https://github.com/catalystneuro/nwb-conversion-tools/pull/181/commits/d25494a34bcc8fc0dacb873a416482269e5ec1c8)that dont require checking for existence of 'description' as json-schema would take care of it. Checking for 'data' is still there though.